### PR TITLE
Properly handle joins and propagate invalidation assessment

### PR DESCRIPTION
**Pull request by ahuang11**
I am now properly able to back out of the ephemeral table and consistently get the desired result.

image

image

image

The root issue was that if there were multiple tables within a single source that needed to be joined, the key `a_source_name` was overridden, so no joins really happened.
`sources[a_source_name] = (a_source, a_table)`

I refactored it so the table name is the key; and if there are duplicate it prefixes with the source name and then re-extracts it out when fetching the table.

### CODE REVIEW

File: lumen/ai/agents.py [hunk 1]
```diff
@@ -376,7 +379,7 @@ async def answer(self, messages: list | str):
             for source in available_sources:
                 for table in source.get_tables():
                     tables_to_source[table] = source
-                    schema = get_schema(source, table, include_min_max=False, include_enum=False, limit=1)
+                    schema = get_schema(source, table, include_min_max=False, include_enum=True, limit=1)
```
Comment (ahuang11) [inline lumen/ai/agents.py:382]: An enum can give a bit more context, especially if the column names are ambiguous.

image

File: lumen/ai/agents.py [hunk 2]
```diff
@@ -363,9 +363,12 @@ class TableAgent(LumenBaseAgent):
 
     @staticmethod
     def _create_table_model(tables):
-        table_model = create_model("Table", table=(Literal[tables], FieldInfo(
-            description="The most relevant table based on the user query; if none are relevant, select the first."
-        )))
+        table_model = create_model(
+            "Table",
+            relevant_table=(Literal[tables], FieldInfo(
```
Comment (ahuang11) [inline lumen/ai/agents.py:368]: Being more explicit helps LLM hone in on what it has to do.

File: lumen/ai/agents.py [hunk 3]
```diff
@@ -542,20 +545,24 @@ async def _create_valid_sql(self, messages, system, sources, errors=None):
         if not sql_query:
             raise ValueError("No SQL query was generated.")
 
+        sources = set(tables_to_source.values())
         if len(sources) > 1:
             mirrors = {}
-            for a_source, a_table in sources.values():
+            for a_table, a_source in tables_to_source.items():
```
Comment (ahuang11) [inline lumen/ai/agents.py:551]: Basically reversed it and dropped one layer; I don't think the keys initially were useful.

File: lumen/ai/agents.py [hunk 4]
```diff
@@ -599,50 +606,8 @@ async def _create_valid_sql(self, messages, system, sources, errors=None):
...
-        return sources
-
-    async def answer(self, messages: list | str):
-        source = memory["current_source"]
-        table = memory["current_table"]
-
-        if not hasattr(source, "get_sql_expr"):
-            return None
-
+    async def check_join_required(self, messages, schema, table):
```
Comment (ahuang11) [inline lumen/ai/agents.py:609]: Simple refactoring into a separate util method to reduce code clutter in `answer`

File: lumen/ai/prompts/check_validity.jinja2 [hunk 5]
```diff
@@ -1,4 +1,8 @@
-Based on the latest user's query, decide whether the current table and data contains the the required data. If the user is referencing "this data" or "that" they probably are referring to the current data. Pay particular attention to whether the data actually contains the columns required to answer the query, e.g. if they are asking for a location but there are no location related column the data is invalid.
+Based on the latest user's query, decide whether the current table and data contains the the required data. If the user
+is referencing "this data" or "that" they probably are referring to the current data. Pay particular attention to
+whether the data actually contains the columns required to answer the query, e.g. if they are asking for a location but
+there are no location related column the data is invalid. However, if the query can be solved through SQL,
+the data is assumed to be valid.
```
Comment (philippjfr) [inline lumen/ai/prompts/check_validity.jinja2:5]: Why add line breaks?
```suggestion
Based on the latest user's query, decide whether the current table and data contains the the required data. If the user is referencing "this data" or "that" they probably are referring to the current data. Pay particular attention to whether the data actually contains the columns required to answer the query, e.g. if they are asking for a location but there are no location related column the data is invalid. However, if the query can be solved through SQL, the data is assumed to be valid.
```

File: lumen/ai/prompts/sql_query.jinja2 [hunk 6]
```diff
@@ -10,6 +10,7 @@ The data for table '{{ table }}' follows the following YAML schema:
 It was already determined that no join is required, so only use the existing
 table '{{ table }}' to calculate the result.
```
Comment (philippjfr) [inline lumen/ai/prompts/sql_query.jinja2:11]:
```suggestion
It was already determined that no join is required, so only use the existing table '{{ table }}' to calculate the result.
```

File: lumen/ai/prompts/sql_query.jinja2 [hunk 7]
```diff
@@ -10,6 +10,7 @@ The data for table '{{ table }}' follows the following YAML schema:
...
 {% else %}
+Please perform a join between the necessary tables.
 If the join's values do not align based on the min/max lowest common denominator, then perform a join based on the
 closest match, or resample and aggregate the data to align the values.
```
Comment (philippjfr) [inline lumen/ai/prompts/sql_query.jinja2:15]:
```suggestion
If the join's values do not align based on the min/max lowest common denominator, then perform a join based on the closest match, or resample and aggregate the data to align the values.
```

**Review by philippjfr — APPROVED** [review]

**Pull request merged**